packages/typespec-client-generator-core/src/validate.ts

```diff
@@ -1,12 +1,52 @@
-import { Program } from "@typespec/compiler";
-import { createTCGCContext } from "./internal-utils.js";
-import { getAllModelsWithDiagnostics } from "./types.js";
+import { Enum, listServices, Model, Program } from "@typespec/compiler";
+import { AllScopes, getClientNameOverride } from "./decorators.js";
+import { createTCGCContext, DuplicateTracker, TCGCContext } from "./internal-utils.js";
+import { createStateSymbol, reportDiagnostic } from "./lib.js";
 
 export function $onValidate(program: Program) {
-  // Pass along any diagnostics that might be returned from the HTTP library
   const tcgcContext = createTCGCContext(program);
-  const [_, diagnostics] = getAllModelsWithDiagnostics(tcgcContext);
-  if (diagnostics.length > 0) {
-    program.reportDiagnostics(diagnostics);
+  const languageScopes = getDefinedLanguageScopes(program);
+  validateModels(program, tcgcContext, languageScopes);
+}
+
+function getDefinedLanguageScopes(program: Program): Set<string | symbol> {
+  const languageScopes = new Set<string | symbol>();
+  const stateMap = program.stateMap(createStateSymbol("clientName"));
```

Comment:
I would recommend not getting the data here but keeping accesor to this state with types in the same place. Just makes sure you don't have 5 different places that assume the type of the state.

Reply:
I don't quite follow here, I need to get the `clientName` state map in order to get the defined language scopes within it.
Are you saying I should only get the `clientName` state map in a central place and get defined scopes prepared in advance, and pass that along to this function instead?

Reply:
yeah meant that, only have a single place that create the state symbol and provide typed function to use that elsewhere

Reply:
moved `clientNameKey` to `internal-utils`